plugins/syntax/number.ml

```diff
@@ -0,0 +1,479 @@
+     These checks are neither sound nor complete, so we only warn. *)
+  let () =
+    (* associate inductives to types, and check that this mapping is one to one
+       and maps [ty_ind] to [ty_name] *)
+    let ind2ty, ty2ind =
+      let add loc ckey cval m =
+        match CMap.find_opt ckey m with
+        | None -> CMap.add ckey cval m
+        | Some old_cval ->
+           if not (Constr.equal old_cval cval) then
+             warn_via_remapping ?loc (env, sigma, ckey, old_cval, cval);
+           m in
+      List.fold_left
+        (fun (ind2ty, ty2ind) (lcnst, _, (_, tcnst), lindc, _, (_, tindc), _) ->
+          add lcnst tindc tcnst ind2ty, add lindc tcnst tindc ty2ind)
+        CMap.(singleton ty_ind ty_name, singleton ty_name ty_ind) l in
+    (* check that type of constants and constructors mapped in [l]
+       match modulo [ind2ty] *)
+    let rm_impls impls (l, t) =
+      let rec aux impls l = match impls, l with
+        | Some _ :: impls, _ :: b -> aux impls b
+        | None :: impls, (n, a) :: b -> (n, a) :: aux impls b
+        | _ -> l in
+      aux impls l, t in
+    let replace m (l, t) =
+      let apply_m c = try CMap.find c m with Not_found -> c in
+      List.fold_right (fun (na, a) b -> Constr.mkProd (na, (apply_m a), b))
+        l (apply_m t) in
+    List.iter (fun (_, cnst, tcnst, loc, indc, tindc, impls) ->
+        let tcnst = rm_impls impls tcnst in
+        let tcnst' = replace CMap.empty tcnst in
+        if not (Constr.equal tcnst' (replace ind2ty tindc)) then
+          let actual = replace CMap.empty tindc in
+          let expected = replace ty2ind tcnst in
+          warn_via_type_mismatch ?loc (env, sigma, indc, cnst, expected, actual))
+      l in
+  (* Associate an index to each inductive, starting from 0 for [ty_ind]. *)
+  let ind2num, num2ind, nb_ind =
+    CMap.fold (fun ind _ (ind2num, num2ind, i) ->
+        CMap.add ind i ind2num, Int.Map.add i ind num2ind, i + 1)
+      (CMap.remove ty_ind constructors)
+      (CMap.singleton ty_ind 0, Int.Map.singleton 0 ty_ind, 1) in
+  (* Finally elaborate [to_post] *)
+  let to_post =
+    let rec map_prod impls tindc = match impls with
+      | Some _ :: impls -> ToPostHole :: map_prod impls tindc
+      | _ ->
+         match tindc with
+         | [] -> []
+         | (_, a) :: b ->
+            let t = match CMap.find_opt a ind2num with
+              | Some i -> ToPostAs i
+              | None -> ToPostCopy in
+            let impls = match impls with [] -> [] | _ :: t -> t in
+            t :: map_prod impls b in
+    Array.init nb_ind (fun i ->
+        List.map (fun indc ->
+            let _, cnst, _, _, _, tindc, impls = List.find (eq_indc indc) l in
+            indc, cnst, map_prod impls (fst tindc))
+          (CMap.find (Int.Map.find i num2ind) constructors)) in
+  (* and use constants mapped to constructors of [ty_ind] as triggers. *)
+  let pt_refs = List.map (fun (_, cnst, _) -> cnst) (to_post.(0)) in
```

Comment:
Note that this doesn't quite work if, e.g., you want to support both `list byte` and `list ascii`.  Perhaps this is an extension for the future, though it would be good to document this limitation.

Reply:
Hum, looks like another bug (c.f. next comment).